packages/components/src/query-controls/terms.ts

```diff
@@ -22,22 +22,29 @@ import type {
  */
 export function buildTermsTree( flatTerms: readonly ( Author | Category )[] ) {
 	const flatTermsWithParentAndChildren: TermWithParentAndChildren[] =
-		flatTerms.map( ( term ) => {
-			return {
-				children: [],
-				parent: null,
-				...term,
-				id: String( term.id ),
-			};
-		} );
+		flatTerms.map( ( term ) => ( {
+			children: [],
+			parent: null,
+			...term,
+			id: String( term.id ),
+		} ) );
 
-	const termsByParent: TermsByParent = groupBy(
-		flatTermsWithParentAndChildren,
-		'parent'
-	);
-	if ( termsByParent.null && termsByParent.null.length ) {
+	if ( ! ensureParents( flatTermsWithParentAndChildren ) ) {
```

Comment:
I wonder if we could just inline `ensureParents()`. For me it was easier to read this line:

```
terms.every( ( term ) => term.parent !== null )
```

than having to read what `ensureParents()` does. It's not immediately clear what the function does too - does it add a parent to each term? Or does it filter only terms with parents? I wouldn't expect from a function with this name to return `true` only when ALL passed terms have parents. Inlining the condition will solve this mystery IMHO.

Reply:
Inlining didn't work for me because TypeScript doesn't infer that `parent` isn't `null` anymore. That's why I created a type guard. That's the important part:

> `terms is ( TermWithParentAndChildren & { parent: number } )[]` (L13)

Reply:
I see. Let's think of a name that matches the purpose better then.

Reply:
I have an idea — what if we tweaked the way the `termsByParent` dictionary is created, by literally using the `'null'` string as a fallback key when the `parent` is actually `null` ?

<details>
<summary>Something like this</summary>

diff --git a/packages/components/src/query-controls/terms.ts b/packages/components/src/query-controls/terms.ts
index 951ca626be..9f5688fc91 100644
--- a/packages/components/src/query-controls/terms.ts
+++ b/packages/components/src/query-controls/terms.ts
@@ -8,11 +8,6 @@ import type {
 	TermsByParent,
 } from './types';
 
-const ensureParents = (
-	terms: TermWithParentAndChildren[]
-): terms is ( TermWithParentAndChildren & { parent: number } )[] => {
-	return terms.every( ( term ) => term.parent !== null );
-};
 /**
  * Returns terms in a tree form.
  *
@@ -29,22 +24,23 @@ export function buildTermsTree( flatTerms: readonly ( Author | Category )[] ) {
 			id: String( term.id ),
 		} ) );
 
-	if ( ! ensureParents( flatTermsWithParentAndChildren ) ) {
-		return flatTermsWithParentAndChildren;
-	}
-
 	const termsByParent = flatTermsWithParentAndChildren.reduce(
 		( acc: TermsByParent, term ) => {
 			const { parent } = term;
-			if ( ! acc[ parent ] ) {
-				acc[ parent ] = [];
+			const parentKey = parent ?? 'null';
+			if ( ! acc[ parentKey ] ) {
+				acc[ parentKey ] = [];
 			}
-			acc[ parent ].push( term );
+			acc[ parentKey ].push( term );
 			return acc;
 		},
 		{}
 	);
 
+	if ( termsByParent.null && termsByParent.null.length ) {
+		return flatTermsWithParentAndChildren;
+	}
+
 	const fillWithChildren = (
 		terms: TermWithParentAndChildren[]
 	): TermWithParentAndChildren[] => {

</details>

We could even decide to use another string instead of `'null'` — there shouldn't be any risk of conflict with other dictionary keys, since`parent` should be a `number`

Reply:
I do think that's a good use case for `.every` (or `.some`) and drop out early before constructing the object and then throwing it away. I'll have another look and see what I come up with.

Reply:
Sounds good. It looks like the `buildTermsTree` already has some unit tests, so we should have good confidence in applying the changes you suggested!

Reply:
In https://github.com/WordPress/gutenberg/pull/48779/commits/a3745b0f8ce3d082cb511bfd832676110518f480 I tried to find a better name for the type guard and added a comment about why we're doing it. I think that should address the original concern and please let me know if there's any other feedback!

Reply:
I'd name the guard function `areAllParentsDefined` or `allTermsHaveParents` but then I'd leave it to your judgment.